src/modules/user-auth/utils/AuthGuard/AuthGuard.tsx

```diff
@@ -0,0 +1,38 @@
+import { FC, ReactNode, useEffect } from 'react';
+import { useAuth } from '@modules/user-auth/contexts/AuthContext';
+import DefaultUnauthenticatedPrompt from '../../components/DefaultUnauthenticatedPrompt';
+import DefaultEmailUnverifiedPrompt from '../../components/DefaultEmailUnverifiedPrompt';
+import { useLayoutModal } from '@modules/layouts/contexts/LayoutModalContext';
+
+type AuthGuardProps = {
+  emailVerified?: boolean;
+  promptLogin?: boolean;
+  renderUnauthenticated?: ReactNode;
+  renderEmailUnverified?: ReactNode;
+};
+
+const AuthGuard: FC<AuthGuardProps> = ({
+  children,
+  emailVerified = false,
+  promptLogin = false,
+  renderUnauthenticated = <DefaultUnauthenticatedPrompt />,
+  renderEmailUnverified = <DefaultEmailUnverifiedPrompt />,
+}) => {
+  const { ready, user } = useAuth();
+  const { loginModal } = useLayoutModal();
+
+  useEffect(() => {
+    if (promptLogin && ready && !user) {
+      loginModal.show();
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [ready, user]);
+
+  if (!user) return <>{renderUnauthenticated}</>;
+
+  if (emailVerified && !user.emailVerified) return <>{renderEmailUnverified}</>;
```

Comment:
Is `emailVerified` necessary? We should always show the message when the user's email is not verified right?

Reply:
Not really, for Account/Profile related settings, we should always allow access without a verified email (there're some laws about information accessibility that we have to allow the ease of accessing or something like that).

Again, we can change the default value but it's better to leave the `props` there for flexbility.